### ISSUE TITLE
fl_choice() doesn't tell you if the dialog was closed

### ISSUE THREAD

**Issue by CendioOssman** [fltk/fltk#282]
Right now the return value of `0` from `fl_choice()` is a bit ambiguous as it can mean two different things:

 * The first button was pressed
 * The dialog was closed (e.g. using Escape)

Essentially this forces you to have the first button be the default "cancel" operation, which can be a bit inflexible.

Not sure how to solve this without breaking backwards compatibility, but ideally I'd like different return codes for these two cases. E.g. `-1` if the window is closed.

**Comment by Albrecht-S** [comment]
Would it help to add a function like `fl_message_window_closed()` that would **return 1 if the last dialog** (fl_ask, fl_choice, etc.) was closed w/o pressing a button (or using its shortcut), i.e. by Escape or the window close button? User code would then look like:
```
  int ret = fl_choice(...);
  if ( /* ret == 0 && */ fl_message_window_closed())
    ret = -1;
```
You could put that in your own `fl_choice()` wrapper or use it whenever you need it. I don't think you need the "`ret == 0 &&`" part but I added it above for clarity.

I also thought about making it possible to assign a callback for [ the next call of ] a standard dialog which would get the window close event (by Escape or window close button) but that seems to be more complicated (a) for the user to set up and (b) for the internal implementation. Or maybe not.

Thoughts, other ideas?

PS: In the future I'd like to get away from these static dialog windows and functions. We need a proper dialog class that can be called by the user or conveniency functions like fl_choice() etc. and can have multiple instances at a time [1]. This new dialog class could take care of this. But that's not yet available.

[1] Multiple instances can be necessary if such a dialog is triggered by a timeout while another dialog is still open. See also [STR 3242: Re-enable nested (aka "recursive") common dialogs](https://www.fltk.org/str.php?L3242). The title is somewhat misleading, it's more about **parallel** than **nested** dialogs.

**Comment by Albrecht-S** [comment]
FTR: I've implemented something as described above in my attempt to fix STR 3242 and this issue together. Tests look good but I'll need some more tests and I'm now working on improving the docs. That said, ...

@CendioOssman Would you also like to distinguish closing the window with the `Escape` key or the window close button?

Another option would be to use a new function name like `fl_choice2` that returns e.g. `-1` for `Escape` and `-2` for window close (or vice versa). Both new return values would be negative to simplify handling so you can use `ret <= 0` to check if one of these occurred.

**Comment by rempas** [comment]
> Backwards compatibility is a very high value for a library like FLTK. We're not going to break existing programs by a new version of the library.

Oh! That's a great goal! Then whatever you do will come in handy ;)

**Comment by Albrecht-S** [comment]
> A little clunky, but acceptable if that's the best way to avoid ABI/API changes. :)

@CendioOssman Thanks for your comment. My current implementation for 1.4.0 avoids this extra call by adding a new function as I mentioned above. I'm not yet clear about the name though. `fl_choice2()` would be one option, `fl_choice_ex()` or `fl_choice_ext()` are other options. I'm (still) open for suggestions.

Due to the massive changes in the implementation caused by [STR 3242](https://www.fltk.org/str.php?L3242) it's very unlikely that this will ever be available in 1.3 though (we don't add new features anymore).

> I can't think of a use case for that off hand, no.

The new function will return `-1` for window close, no matter if the `Escape` key was hit or the window close button was pushed.

Your function name choice or "better" proposals would be appreciated.

**Comment by CendioOssman** [comment]
Isn't 1.4 an API break so you can just modify the existing function? I was only thinking of the 1.3 branch for backwards compatibility.

**Comment by Albrecht-S** [comment]
> 1.4 would be for ABI breaks only, AFAIK.

As Greg said. API breaks are bad because they would break existing user programs. We try to avoid that in all 1.x versions, unless ... (see Greg's comment). Adding new functions is always possible, this would neither break the API nor the ABI.

Deprecated functions and methods may be removed in later versions which would break the API - but that's something that needs to be done at some time. For instance: deprecated in 1.3: remove in 1.5 or later. But we're **very** conservative WRT removal of deprecated functions.

**Comment by Albrecht-S** [comment]
@CendioOssman FTR: I'm now looking for a solution for FLTK 1.3.8 (upcoming release) for this issue although I wrote it would be very unlikely. I'm still not sure that it's doable (50:50).

**Comment by CendioOssman** [comment]
This is not a critical issue. We'd still have to consider how old FLTK behaves for some time anyway.

But for the long term it would be a nice improvement.

**Comment by Albrecht-S** [comment]
Fixed in FLTK 1.3.8, and **_FLTK 1.3.8 has been released today!_**

The fix in FLTK 1.4.0 still needs some minor modifications but will be functionally identical.

**Comment by Albrecht-S** [comment]
Fixed in commit 130f864d1d3f02f854a5f4085a49318a03a8eea0 (branch master) as well.
